services/src/main/java/com/chatbot/services/AsyncServices/HangoutsAsyncService.java

```diff
@@ -0,0 +1,144 @@
+  @Autowired
+  private IDMapping iDMapping;
+  @Autowired
+  private DialogflowConversation dialogflowConversation;
+
+  // handler for a request from the chat client
+  @Override
+  @Async("asyncExecutor")
+  public void chatServiceRequestHandler(final ChatServiceRequest chatServiceRequest)
+      throws Exception {
+    final String spaceID = chatServiceRequest.getSender().getChatClientGeneratedId();
+    switch (chatServiceRequest.getRequestType()) {
+      case ADDED:
+        hangoutsMessageSender.sendMessageBasedOnMessageType(spaceID, THANKS_FOR_ADDING_MESSAGE,
+            HANGOUTS_MESSAGE_TYPE.TEXT);
+        iDMapping.addNewMapping(chatServiceRequest.getSender().getChatClientGeneratedId(),
+            chatServiceRequest.getSender().getUserId(), chatServiceRequest.getChatClient());
+        break;
+      case REMOVED:
+        break;
+      case MESSAGE:
+        handleMessageEvent(chatServiceRequest);
+        break;
+      default:
+        break;
+    }
+  }
+
+  // handle requests which contain a user message
+  public void handleMessageEvent(final ChatServiceRequest chatServiceRequest) throws Exception {
+    final String spaceID = chatServiceRequest.getSender().getChatClientGeneratedId();
+    // The spaceID of the user is used as the sessionID for hangouts
+    dialogflowConversation.setSessionID(spaceID);
+    if (chatServiceRequest.getUserMessage().getAttachmentsCount() == 0) {
+      final Value userID = Value.newBuilder()
+          .setStringValue(chatServiceRequest.getSender().getUserId()).build();
+      final Struct payload = Struct.newBuilder()
+          .putFields("userID", userID)
+          .build();
+      final QueryResult queryResult = dialogflowConversation.sendMessage(
+          chatServiceRequest.getUserMessage().getText(), payload);
+      final String response = queryResult.getFulfillmentText();
+      if(ChatServiceConstants.LIST_OF_INTENTS_WITH_INTERACTIVE_RESPONSE
+          .contains(queryResult.getIntent().getDisplayName())) {
+        hangoutsMessageSender.sendMessageBasedOnMessageType(spaceID, response,
+            HANGOUTS_MESSAGE_TYPE.CARD_WITH_HEADER);  
+      } else {
+        hangoutsMessageSender.sendMessageBasedOnMessageType(spaceID, response,
+            HANGOUTS_MESSAGE_TYPE.TEXT);
+      }
+    } else {
+      final List<String> currentContextList = dialogflowConversation.getCurrentContexts();
+      if (currentContextList.contains(ChatServiceConstants.EXPECTING_IMAGES_CONTEXT)) {
+        // TODO: send images to backend
+        hangoutsMessageSender.sendMessageBasedOnMessageType(
+            chatServiceRequest.getSender().getChatClientGeneratedId(), IMAGES_RECEIVED_MESSAGE,
+            HANGOUTS_MESSAGE_TYPE.TEXT);
+      } else {
+        hangoutsMessageSender.sendMessageBasedOnMessageType(
+            chatServiceRequest.getSender().getChatClientGeneratedId(), NOT_EXPECTING_IMAGE_MESSAGE,
+            HANGOUTS_MESSAGE_TYPE.TEXT);
+      }
```

Comment:
Can be extracted in a private method as used multiple times

Reply:
Only used here, right?